src/Microsoft.AspNet.Security.OpenIDConnect/OpenidConnectAuthenticationHandler.cs

```diff
@@ -0,0 +1,621 @@
+                ResponseMode = OpenIdConnectResponseModes.FormPost,
+                ResponseType = Options.ResponseType,
+                Scope = Options.Scope,
+                State = OpenIdConnectAuthenticationDefaults.AuthenticationPropertiesKey + "=" + Uri.EscapeDataString(Options.StateDataFormat.Protect(properties))
+            };
+
+            // TODO - brentschmaltz, if INonceCache is set should we even consider if ProtocolValidator is set?
+            if (Options.ProtocolValidator.RequireNonce)
+            {
+                openIdConnectMessage.Nonce = Options.ProtocolValidator.GenerateNonce();
+                if (Options.NoneCache != null)
+                {
+                    Options.NoneCache.AddNonce(openIdConnectMessage.Nonce);
+                }
+                else
+                {
+                    AddNonce(openIdConnectMessage.Nonce);
+                }
+            }
+
+            var notification = new RedirectToIdentityProviderNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options)
+            {
+                ProtocolMessage = openIdConnectMessage
+            };
+
+            await Options.Notifications.RedirectToIdentityProvider(notification);
+            if (!notification.HandledResponse)
+            {
+                string redirectUri = notification.ProtocolMessage.CreateAuthenticationRequestUrl();
+                if (!Uri.IsWellFormedUriString(redirectUri, UriKind.Absolute))
+                {
+                    _logger.WriteWarning("The authenticate redirect URI is malformed: " + redirectUri);
+                }
+
+                Response.Redirect(redirectUri);
+            }
+        }
+
+        protected override AuthenticationTicket AuthenticateCore()
+        {
+            return AuthenticateCoreAsync().GetAwaiter().GetResult();
+        }
+
+        /// <summary>
+        /// Invoked to process incoming authentication messages.
+        /// </summary>
+        /// <returns>An <see cref="AuthenticationTicket"/> if successful.</returns>
+        protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
+        {
+            // Allow login to be constrained to a specific path. Need to make this runtime configurable.
+            if (Options.CallbackPath.HasValue && Options.CallbackPath != (Request.PathBase + Request.Path))
+            {
+                return null;
+            }
+
+            OpenIdConnectMessage openIdConnectMessage = null;
+
+            // assumption: if the ContentType is "application/x-www-form-urlencoded" it should be safe to read as it is small.
+            if (string.Equals(Request.Method, "POST", StringComparison.OrdinalIgnoreCase)
+              && !string.IsNullOrWhiteSpace(Request.ContentType)
+                // May have media/type; charset=utf-8, allow partial match.
+              && Request.ContentType.StartsWith("application/x-www-form-urlencoded", StringComparison.OrdinalIgnoreCase)
```

Comment:
FYI: I'm redoing the GetFormAsync right now so we'll be able to remove much of this code.